### PR TITLE
Ignore loopback and private network addresses (such as 192.168.0.1)

### LINKED ISSUE #18474: Exclude all local subnets from HTTPSE
[//]: # (Thank you for reporting an issue to HTTPS Everywhere.)
[//]: # (We welcome input from users on improving this project.)
[//]: #
[//]: # (Please help us by following this issue template.)
[//]: # (You can delete all blank lines and all lines starting)
[//]: # (with the comment marker, such as this one.)

Type: code issue

Coming in from email:

> Hello,
> I would like to know if HTTPS everywhere can be disabled globally for private ip addresses (10.0.0.0/8 172.16.0.0/12 192.168.0.0/16) since they don't need https most of the time being 
in the same subnet and it would be very inconvenient to disable HTTPS everywhere for every single ip address.

What do people think about this request?

### PR DESCRIPTION
**Pull request by pipboy96**
Closes #18474.

### CODE REVIEW

File: chromium/background-scripts/background.js [hunk 1]
```diff
@@ -327,9 +336,9 @@ function onBeforeRequest(details) {
     (uri.protocol === 'http:' || uri.protocol === 'ftp:') &&
     uri.hostname.slice(-6) !== '.onion' &&
     uri.hostname !== 'localhost' &&
-    !/^127(\.[0-9]{1,3}){3}$/.test(uri.hostname) &&
     uri.hostname !== '0.0.0.0' &&
```
Comment (pipboy96) [inline chromium/background-scripts/background.js]: @zoracon Should this be moved into `ip_utils`?

File: chromium/test/ip_utils_test.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      assert(parseIp('1.1.1.1...') === -1)
+    })
+
+    it('rejects an IP address-like string with omitted parts', () => {
+      assert(parseIp('.1.1.1') === -1)
+      assert(parseIp('1..1.1') === -1)
+      assert(parseIp('1.1..1') === -1)
+      assert(parseIp('1.1.1.') === -1)
+    })
+
```
Comment (cschanaj) [inline chromium/test/ip_utils_test.js]:
```suggestion
    it('rejects an IP address-like string with invalid parts', () => {
      assert(parseIp('192.168.1.256') === -1)
      assert(parseIp('192.168.256.1') === -1)
      assert(parseIp('192.256.1.1') === -1)
      assert(parseIp('256.168.1.1') === -1)
      assert(parseIp('256.168.1.-1') === -1)
    })

```

Comment (cschanaj) [reply chromium/test/ip_utils_test.js]: not sure if this help to improve the coverage to 100%

**Comment by pipboy96** [comment]
@cschanaj I think there is no rule about use of `;`, but @zoracon would be a better person to ask.

**Comment by pipboy96** [comment]
@cschanaj Thanks! It now has 100% coverage.

File: chromium/background-scripts/background.js (hunk 1 above)
Comment (zoracon) [reply chromium/background-scripts/background.js]: Go ahead and move that into `ip_utils`

**Comment by zoracon** [comment]
> @cschanaj I think there is no rule about use of `;`, but @zoracon would be a better person to ask.

@pipboy96 let's stick to using semicolons since I want as much consistency and readability as we can in the repo. There was no hard rule in place but it is general practice. The file is generally clean but want to avoid future potential debugging issues like concatenated statements.

**Comment by pipboy96** [comment]
@zoracon any idea what causes this error?

**Comment by zoracon** [comment]
@pipboy96 Merged in that web-crypto temporary fix. Will investigate more

**Comment by taokann** [comment]
Hello, I searched the web for this issue and found here that's already resolved.
Unfortunately I have the right version (2019.11.7) but I still have to manually whitelist all local IPs.
I read the PR's code, I don't know the project well but it seems that the feature is only added in the Chromium version. Is it right ? Will it be available for Firefox too ? I'm using Firefox 68.3.0 ESR with Debian Buster.

**Comment by pipboy96** [comment]
@taokann Please wait until the next HTTPS Everywhere extension release, which would include #18755 patch that fixes a typo in this PR.

**Pull request merged**